edb/pgsql/schemamech.py

```diff
@@ -50,361 +53,412 @@
+
+def compile_constraint(
+    subject: s_types.Type,
+    constraint: s_constraints.Constraint,
+    schema: s_schema.Schema,
+    source_context: Optional[parsing.ParserContext],
+) -> SchemaDomainConstraint | SchemaTableConstraint:
+    assert constraint.get_subject(schema) is not None
+    TypeOrPointer = s_types.Type | s_pointers.Pointer
+
+    constraint_origins = constraint.get_constraint_origins(schema)
+    first_subject = constraint_origins[0].get_subject(schema)
+
+    is_optional = isinstance(
+        first_subject, s_pointers.Pointer
+    ) and not first_subject.get_required(schema)
+    singletons: Collection[Tuple[TypeOrPointer, bool]] = frozenset(
+        {(subject, is_optional)}
+    )
+    options = qlcompiler.CompilerOptions(
+        anchors={qlast.Subject().name: subject},
+        path_prefix_anchor=qlast.Subject().name,
+        apply_query_rewrites=False,
+        singletons=singletons,
+        schema_object_context=type(constraint),
+        # Remap the constraint origin to the subject, so that if
+        # we have B <: A, and the constraint references A.foo, it
+        # gets rewritten in the subtype to B.foo. It's OK to only
+        # look at one constraint origin, because if there were
+        # multiple different origins, they couldn't get away with
+        # referring to the type explicitly.
+        type_remaps={first_subject: subject},
+    )
+
+    final_expr = constraint.get_finalexpr(schema)
+    assert final_expr is not None and final_expr.qlast is not None
+    ir = qlcompiler.compile_ast_to_ir(
+        final_expr.qlast,
+        schema,
+        options=options,
+    )
+    assert isinstance(ir, irast.Statement)
+    assert isinstance(ir.expr.expr, irast.SelectStmt)
+
+    except_data = None
+    if except_expr := constraint.get_except_expr(schema):
+        except_ir = qlcompiler.compile_ast_to_ir(
+            except_expr.qlast,
+            schema,
+            options=options,
+        )
+        except_sql, _ = compiler.compile_ir_to_sql_tree(
+            except_ir, singleton_mode=True
+        )
+        except_data = _edgeql_tree_to_expr_data(except_sql)
+
+    terminal_refs = ir_utils.get_longest_paths(ir.expr.expr.result)
+    ref_tables = get_ref_storage_info(ir.schema, terminal_refs)
+
+    if len(ref_tables) > 1:
+        raise errors.InvalidConstraintDefinitionError(
+            f'Constraint {constraint.get_displayname(schema)} on '
+            f'{subject.get_displayname(schema)} is not supported '
+            f'because it would depend on multiple objects',
+            context=source_context,
+        )
+    elif ref_tables:
+        subject_db_name, info = next(iter(ref_tables.items()))
+        table_type = info[0][3].table_type
+    else:
+        # the expression does don't have any refs: default to the subject table
+
+        if isinstance(subject, s_pointers.Pointer):
+            subject_table = subject.get_source(schema)
+        else:
+            subject_table = subject
```

Comment:
This is the only impactful change. Should I put it into a separate PR?

Reply:
(or just make sure it's a separate commit)

Reply:
I've squashed it into two commits that will be "rebase and merge"d.

Reply:
Hm, except this wouldn't cherry-pick cleanly without the refactor?  How comfortable are we with cherry-picking the refactor?